src/levanter/eval_harness.py

```diff
@@ -0,0 +1,564 @@
+
+    def to_task_dict(self) -> dict:
+        import lm_eval.tasks as tasks
+
+        manager = tasks.TaskManager()
+        # we need to do it this way b/c i can't figure out how to run e.g. hellaswag 0 shot and 10 shot in a single run
+        this_tasks = {}
+        for task in self.task_spec_or_default():
+            try:
+                if isinstance(task, str):
+                    this_tasks.update(tasks.get_task_dict(task, manager))
+                else:
+                    our_name = task.get("task_alias", task["task"]) if isinstance(task, dict) else task
+                    our_name = our_name.replace(" ", "_")
+                    task_dict = tasks.get_task_dict([task], manager)
+                    this_task = task_dict.popitem()[1]
+                    # hacky, but this allows us to run multiple instances of the same task with different fewshot settings
+                    this_task.config.task = our_name
+                    this_tasks[our_name] = this_task
+            except Exception:
+                logger.exception(f"Failed to load task {task}")
+                raise ValueError(f"Failed to load task {task}")
+        return this_tasks
+
+
+@dataclass(frozen=True)
+class EvalHarnessMainConfig:
+    tokenizer: str
+    checkpoint_path: str
+    checkpoint_is_hf: bool = False
+    trainer: TrainerConfig = dataclasses.field(default_factory=TrainerConfig)
+    model: LmConfig = dataclasses.field(default_factory=Gpt2Config)
+
+    eval_harness: LmEvalHarnessConfig = dataclasses.field(default_factory=LmEvalHarnessConfig)
+
+    @property
+    def EvalBatch(self):
+        return self.trainer.EvalBatch
+
+    @cached_property
+    def the_tokenizer(self):
+        return load_tokenizer(self.tokenizer)
+
+
+def run_lm_eval_harness(
+    config: LmEvalHarnessConfig,
+    model,
+    tokenizer,
+    EvalBatch,
+    axis_resources,
+) -> dict:
+    # tasks_to_run = tasks.get_task_dict(config.task_spec_or_default(), tasks.TaskManager())
+    tasks_to_run = config.to_task_dict()
+
+    outputs = _actually_run_eval_harness(config, model, tasks_to_run, tokenizer, EvalBatch, axis_resources)
+
+    return outputs
+
+
+def _actually_run_eval_harness(config: LmEvalHarnessConfig, model, tasks_to_run, tokenizer, EvalBatch, axis_resources):
+    max_examples = config.max_examples
+    max_eval_length = config.max_eval_length
+
+    EvalPos = model.Pos if max_eval_length is None else model.Pos.resize(max_eval_length)
+    harness = LevanterHarnessLM(EvalBatch, EvalPos, model, axis_resources, tokenizer)
+    # we always log_samples here and filter out the samples later if we don't want them
+    outputs = evaluator.evaluate(harness, tasks_to_run, limit=max_examples, log_samples=True)
+
+    averages = _compute_averages(outputs)
+    outputs["averages"] = averages
+
+    if not config.log_samples:
+        del outputs["samples"]
+
+    return outputs
+
+
+def _compute_averages(outputs):
```

Comment:
Unrelated to this PR specifically but I think we could add more ways to aggregate results- the DCLM paper for example reports centered accuracy. Maybe we can have this aggregation function be something we pass intoLmEvalHarnessConfig. Or something like subtract_random_baseline: true in the YAML config...

Reply:
imho we should modify lm harness to do that, but it's a great idea